client/state/posts/reducer.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import indexBy from 'lodash/collection/indexBy';
+
+/**
+ * Internal dependencies
+ */
+import {
+	POST_RECEIVE,
+	POSTS_RECEIVE,
+	POSTS_REQUEST,
+	POSTS_REQUEST_SUCCESS,
+	POSTS_REQUEST_FAILURE
+} from 'state/action-types';
+
+/**
+ * Tracks all known post objects, indexed by post global ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function items( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_RECEIVE:
+			state = Object.assign( {}, state, {
+				[ action.post.global_ID ]: action.post
+			} );
+			break;
+
+		case POSTS_RECEIVE:
+			state = Object.assign( {}, state, indexBy( action.posts, 'global_ID' ) );
+			break;
+	}
+
+	return state;
+}
+
+/**
+ * Returns the updated site posts state after an action has been dispatched.
+ * The state reflects a mapping of site ID, post ID pairing to global post ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function sitePosts( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_RECEIVE:
+		case POSTS_RECEIVE:
+			let posts = action.post ? [ action.post ] : action.posts;
+			posts.forEach( ( post ) => {
+				state = Object.assign( {}, state );
+				if ( ! state[ post.site_ID ] ) {
+					state[ post.site_ID ] = {};
+				}
+
+				state[ post.site_ID ][ post.ID ] = post.global_ID;
+			} );
+			break;
+	}
+
+	return state;
+}
+
+/**
+ * Returns the updated post query state after an action has been dispatched.
+ * The state reflects a mapping of site ID to active queries.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function siteQueries( state = {}, action ) {
+	switch ( action.type ) {
+		case POSTS_REQUEST:
+		case POSTS_REQUEST_SUCCESS:
+		case POSTS_REQUEST_FAILURE:
+			const { siteId, posts } = action;
+			const query = JSON.stringify( action.query );
+
+			state = Object.assign( {}, state );
+			if ( ! state[ siteId ] ) {
+				state[ siteId ] = {};
+			}
+
+			state[ siteId ][ query ] = {
+				fetching: POSTS_REQUEST === action.type
+			};
+
+			if ( posts ) {
```

Comment:
if we have no posts then the site doesn't end up getting a `posts` property, but if we initialized `posts` then we could always end up with an utterable list (maybe an empty list) and remove this conditional…

```js
const { siteId, posts = [] } = action;
```